server/src/test/java/org/elasticsearch/cluster/routing/BatchedRerouteServiceTests.java

```diff
@@ -229,6 +235,93 @@ public void testNotifiesOnFailure() throws InterruptedException {
             }
         }
 
-        assertTrue(countDownLatch.await(10, TimeUnit.SECONDS)); // i.e. it doesn't leak any listeners
+        safeAwait(countDownLatch); // i.e. it doesn't leak any listeners
+    }
+
+    @TestLogging(reason = "testing log output", value = "org.elasticsearch.cluster.routing.BatchedRerouteService:DEBUG")
+    public void testExceptionFidelity() {
+
+        final var mockLogAppender = new MockLogAppender();
+        try (var ignored = mockLogAppender.capturing(BatchedRerouteService.class)) {
+
+            clusterService.getMasterService()
```

Comment:
Might be nice to "separate" the sub test cases with comments. Something like this here and below.

`// First case: assert simulated exception appears unwrapped`